runtimes/core/src/sqldb/manager.rs

```diff
@@ -15,26 +15,48 @@ use crate::trace::Tracer;
 pub struct Manager {
     databases: Arc<HashMap<EncoreName, Arc<DatabaseImpl>>>,
     proxy_port: u16,
+    listener: Mutex<Option<std::net::TcpListener>>,
+    runtime: tokio::runtime::Handle,
 }
 
-impl Manager {
-    pub fn new(
-        clusters: Vec<pb::SqlCluster>,
-        creds: &pb::infrastructure::Credentials,
-        secrets: &secrets::Manager,
-        proxy_port: u16,
-        tracer: Tracer,
-    ) -> anyhow::Result<Self> {
-        let databases = databases_from_cfg(clusters, creds, secrets, proxy_port, tracer)
-            .context("failed to parse SQL clusters")?;
+pub struct ManagerConfig<'a> {
+    pub clusters: Vec<pb::SqlCluster>,
+    pub creds: &'a pb::infrastructure::Credentials,
+    pub secrets: &'a secrets::Manager,
+    pub tracer: Tracer,
+    pub runtime: tokio::runtime::Handle,
+}
+
+impl<'a> ManagerConfig<'a> {
+    pub fn build(self) -> anyhow::Result<Manager> {
+        // Start listening so we can tell which port it is when we generate configuration.
+        let listener =
+            std::net::TcpListener::bind("127.0.0.1:0").context("unable to bind to sqldb port")?;
```

Comment:
will this cause a random port to be selected? or what does `:0` do?

Reply:
Ah found the docs for it https://doc.rust-lang.org/std/net/struct.TcpListener.html#method.bind
TIL